torch/csrc/autograd/autograd_not_implemented_template.h

```diff
@@ -0,0 +1,202 @@
+#pragma once
+
+#include <ATen/core/dispatch/Dispatcher.h>
+#include <torch/csrc/autograd/autograd.h>
+#include <torch/csrc/autograd/VariableTypeUtils.h>
+#include <torch/csrc/autograd/FunctionsManual.h>
+#include <torch/csrc/jit/frontend/function_schema_parser.h>
+
+namespace torch {
+namespace autograd {
+
+namespace detail {
+
+template <typename F>
+struct ForEachTensor : IterArgs<ForEachTensor<F>> {
+  F fn_;
+  mutable size_t idx_arg_ = 0;
+  mutable size_t idx_tensor_ = 0;
+  ForEachTensor(F fn) : fn_(fn) {}
+  void operator()(const variable_list& tensors) const {
+    for (const auto& t : tensors) {
+      fn_(idx_arg_, idx_tensor_, t);
+      idx_tensor_++;
+    }
+    idx_arg_++;
+  }
+  void operator()(const TensorList& tensors) const {
+    for (const auto& t : tensors) {
+      fn_(idx_tensor_, idx_arg_ , t);
+      idx_tensor_++;
+    }
+    idx_arg_++;
+  }
+  void operator()(const c10::optional<at::Tensor>& x) const {
+    // NOLINTNEXTLINE(bugprone-branch-clone)
+    if (x.has_value() && x.value().defined()) {
+      fn_(idx_tensor_, idx_arg_, x.value());
+      idx_tensor_++;
+    }
+    idx_arg_++;
+  }
+  void operator()(const at::Tensor& x) const {
+    fn_(idx_tensor_, idx_arg_, x);
+    idx_tensor_++;
+    idx_arg_++;
+  }
+  template <typename T>
+  void operator()(const T& x) const {
+    // no op
+  }
+};
+
+template <typename F, typename... Args>
+inline void for_each_input_tensor(F fn, Args&&... args) {
+  ForEachTensor<F>(fn).apply(std::forward<Args>(args)...);
+}
+
+// TODO: We do not have access to this because of C++ reasons
+// when building as an extension (why?)- this is available from custom_function.cpp
+bool isFwGradDefined(const c10::optional<Tensor>& t) {
+  return t.has_value() && t->defined() && t->_fw_grad(/*level */ 0).defined();
+}
+
+template <class Mapper, class... Args, size_t... Indices>
+void for_each_tuple(
+    std::tuple<Args...>& tuple,
+    const Mapper& mapper,
+    std::index_sequence<Indices...>) {
+  // We need a context where the parameter pack can be expanded
+  auto l = { (mapper(std::forward<Args>(std::get<Indices>(tuple))), 0)...};
+}
+
+template<typename F>
+void for_each_output_tensor(at::Tensor& t, F fn) {
+  fn(0, 0, t);
+}
+
+template<typename T, typename F>
+void for_each_output_tensor(T& not_tup, F fn) {
+  (ForEachTensor<F>(fn))(not_tup);
+}
+
+template<class... Args, typename F>
+void for_each_output_tensor(std::tuple<Args...>& tup, F fn) {
+  for_each_tuple(tup,  ForEachTensor<F>(fn), std::index_sequence_for<Args...>());
+}
+
+}
+
+template<typename ReturnT, typename... Args>
+struct autogradNotImplementedFallback2 {
```

Comment:
Ideally, we would not land this without also deleting the corresponding code generation logic from the autograd engine. There's not too much, I think, because NotImplemented codegen proceeds very similarly to regular derivative codegen. Maybe what would be cool is if we could generalize this template enough so that it can also work for the NotImplemented case, but that's probably out of scope for this PR.

In any case, this needs to be cross-referenced with the relevant codegen, so when people update the codegen, this template can be updated too (this applies to the boxed fallback too, where code sharing isn't possible). If there are any blocks of logic that can be shared between the three implementations, factoring these out into helper functions would also be a good help.